sklearn/cluster/_k_means_elkan.pyx

```diff
@@ -0,0 +1,176 @@
+        # we could find the closest center in O(n) but
+        # this does not seem to be the bottleneck
+        distance_next_center = np.sort(center_distances, axis=0)[1]
+        if verbose:
+            print("done sorting")
+        for point_index in range(n_samples):
+            upper_bound = upper_bounds[point_index]
+            label = labels[point_index]
+            if distance_next_center[label] >= upper_bound:
+                continue
+            x_p = X_p + point_index * n_features
+            # check other update conditions
+            # TODO: get pointer to lower_bounds[point_index, center_index]
+            for center_index in range(n_clusters):
+                if (center_index != label
+                        and (upper_bound > lower_bounds[point_index, center_index])
+                        and (upper_bound > center_distances[center_index, label])):
+                    # update distance to center
+                    if not bounds_tight[point_index]:
+                        upper_bound = d(x_p, centers_p + label * n_features, n_features)
+                        lower_bounds[point_index, label] = upper_bound
+                        bounds_tight[point_index] = 1
+                    # check for relabels
+                    if (upper_bound > lower_bounds[point_index, center_index]
+                            or (upper_bound > center_distances[label, center_index])):
+                        distance = d(x_p, centers_p + center_index * n_features, n_features)
+                        lower_bounds[point_index, center_index] = distance
+                        if distance < upper_bound:
+                            label = center_index
+                            upper_bound = distance
+            labels[point_index] = label
+            upper_bounds[point_index] = upper_bound
+        if verbose:
+            print("end inner loop")
+        # compute new centers
+        new_centers = _centers_dense(X_, labels_, n_clusters, upper_bounds_)
+        bounds_tight = np.zeros(n_samples, dtype=np.uint8)
+
+        # compute distance each center moved
+        center_shift = np.sqrt(np.sum((centers_ - new_centers) ** 2, axis=1))
+        # update bounds accordingly
+        lower_bounds = np.maximum(lower_bounds - center_shift, 0)
+        upper_bounds = upper_bounds + center_shift[labels_]
+        # reassign centers
+        centers_ = new_centers
+        centers_p = getfloatpointer(new_centers)
+        # update between-center distances
+        center_distances = euclidean_distances(centers_) / 2.
+        if verbose:
+            print('Iteration %i, inertia %s'
+                  % (iteration, np.sum((X_ - centers_[labels]) ** 2)))
+        center_shift_total = np.sum(center_shift)
+        if center_shift_total ** 2 < tol:
+            if verbose:
+                print("center shift %e within tolerance %e"
+                      % (center_shift_total, tol))
+            break
+    if center_shift_total > 0:
```

Comment:
I don't seem to understand why this block of code is here? Why are we recomputing the upper and lower bounds outside the `n_iterations` loop,